File: docs/algorithms/code/algorithms/constructMaximumBinaryTree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function constructMaximumBinaryTree(nums: number[]): TreeNode | null {
  if (!nums.length) {
    return null;
  }

  const rootVal = Math.max(...nums);

  const rootNode = new TreeNode(rootVal);

  const rootIndex = nums.findIndex(item => item === rootVal);

  rootNode.left = constructMaximumBinaryTree(nums.slice(0, rootIndex));
  rootNode.right = constructMaximumBinaryTree(nums.slice(rootIndex + 1));

  return rootNode;
};

File: docs/algorithms/code/algorithms/constructMaximumBinaryTree.go
// 思路：找到最大值，然后递归构造左右子树
// 复杂度分析：时间复杂度 O(n^2)，空间复杂度 O(n)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func constructMaximumBinaryTree(nums []int) *TreeNode {
  if len(nums) == 0 {
    return nil
  }
  max := nums[0]
  maxIndex := 0
  for i := 0; i < len(nums); i++ {
    if nums[i] > max {
      max = nums[i]
      maxIndex = i
    }
  }
  root := &TreeNode{Val: max}
  root.Left = constructMaximumBinaryTree(nums[:maxIndex])
  root.Right = constructMaximumBinaryTree(nums[maxIndex + 1:])
  return root
}